geo/src/algorithm/minimum_rotated_rect.rs

```diff
@@ -41,24 +39,74 @@ where
     type Scalar = T;
 
     fn minimum_rotated_rect(&self) -> Option<Polygon<Self::Scalar>> {
-        let convex_poly = ConvexHull::convex_hull(self);
-        let mut min_area: T = Float::max_value();
-        let mut min_angle: T = T::zero();
-        let mut rect_poly: Option<Polygon<T>> = None;
-        let rotate_point = convex_poly.centroid();
-        for line in convex_poly.exterior().lines_iter() {
-            let (ci, cii) = line.points();
-            let angle = (cii.y() - ci.y()).atan2(cii.x() - ci.x()).to_degrees();
-            let rotated_poly = Rotate::rotate_around_point(&convex_poly, -angle, rotate_point?);
-            let tmp_poly = rotated_poly.bounding_rect()?.to_polygon();
-            let area = tmp_poly.unsigned_area();
+        let hull = ConvexHull::convex_hull(self);
+
+        // We take unit vectors along and perpendicular to each edge, then use
+        // them to build a rotation matrix and apply it to the convex hull,
+        // keeping track of the best AABB found.
+        //
+        // See also the discussion at
+        // https://gis.stackexchange.com/questions/22895/finding-minimum-area-rectangle-for-given-points/22934
+        let mut min_area = <T as Bounded>::max_value();
+        let mut best_edge = None;
+        let (mut best_min_x, mut best_max_x) = (T::zero(), T::zero());
+        let (mut best_min_y, mut best_max_y) = (T::zero(), T::zero());
+        for edge in hull.exterior().lines() {
+            let (dx, dy) = edge.delta().x_y();
+            let norm = dx.hypot(dy);
+            if norm.is_zero() {
+                continue;
+            }
+            let edge = (dx / norm, dy / norm);
+
+            let (mut min_x, mut max_x) = (<T as Bounded>::max_value(), <T as Bounded>::min_value());
+            let (mut min_y, mut max_y) = (<T as Bounded>::max_value(), <T as Bounded>::min_value());
+            for point in hull.exterior().points() {
+                let x = point.y() * edge.1 + point.x() * edge.0;
```

Comment:
[Herbie](https://herbie.uwplse.org) thinks you can improve the accuracy of these two statements using fma:

```rust
let x = point.x().mul_add(edge.0, (point.y() * edge.1));
let y = point.y().mul_add(edge.0, (-point.x() ( edge.1));
```

Does it make any difference?

Reply:
Ooh, good idea, forgot about those.

Reply:
Tried it, slightly better but still no luck :(.

Reply:
Bah.

Reply:
Looks like I can still extract one ULP or so by dividing by `norm` after the multiplications. Reverting FMA still doesn't fix those two vertices.

```
  left: LineString([Coord { x: 1.7000000000000002, y: 24.6 }, Coord { x: 14.649854163628412, y: 25.15341257109523 }, Coord { x: 14.400000000000002, y: 31.000000000000004 }, Coord { x: 1.4501458363715911, y: 30.446587428904774 }, Coord { x: 1.7000000000000002, y: 24.6 }])
 right: LineString([Coord { x: 1.7000000000000004, y: 24.600000000000005 }, Coord { x: 14.64985416362841, y: 25.153412571095235 }, Coord { x: 14.400000000000002, y: 31.000000000000004 }, Coord { x: 1.4501458363715916, y: 30.446587428904774 }, Coord { x: 1.7000000000000004, y: 24.600000000000005 }])
```

Reply:
Sorry to resurrect an old PR, but is herbie something we should consider adding to our CI lint? 

I'd propose it should be *non-failing*, but advisory.

Reply:
If only! The Herbie lint in Rust hasn't worked since sometime in 2016 when the lint framework got rewritten, and nobody has had the energy to resurrect it. I actually asked about it a couple of years ago on URLO and everyone (including one of the Herbie authors) was enthusiastic, but it was clear that I was gonna be on my own. I generated the suggestion in this PR by manually writing it out and feeding it to the online tool.